File: src/infra/location_repository/in_memory.rs
use std::sync::Mutex;

use super::{FetchAllError, LocationRepository};
use crate::domain::models::Location;
use async_trait::async_trait;

pub struct InMemoryLocationRepository {
    error: bool,
    locations: Mutex<Vec<Location>>,
}

impl InMemoryLocationRepository {
    #[cfg(test)]
    pub fn new() -> Self {
        Self {
            error: false,
            locations: Mutex::new(vec![]),
        }
    }

    #[cfg(test)]
    pub fn with_error(self) -> Self {
        Self {
            error: true,
            ..self
        }
    }

    #[cfg(test)]
    pub fn insert(&self, location: Location) {
        self.locations.lock().unwrap().push(location);
    }
}

#[async_trait]
impl LocationRepository for InMemoryLocationRepository {
    async fn fetch_all(&self) -> Result<Vec<Location>, FetchAllError> {
        if self.error {
            return Err(FetchAllError::Unknown("error flag is on"));
        }

        let lock = match self.locations.lock() {
            Ok(lock) => lock,
            _ => return Err(FetchAllError::Unknown("could not acquire lock")),
        };

        Ok(lock.to_vec())
    }
}

File: src/infra/locationRepository/inMemory.ts
import { InMemoryRepository } from "../utils/mod.ts";
import { LocationRepository } from "./mod.ts";
import { Location } from "@domain";
import { Result } from "@utils";

export class InMemoryLocationRepository implements LocationRepository {
    private repo: InMemoryRepository<Location>;

    private constructor(locations: Location[], error: boolean) {
        this.repo = new InMemoryRepository(locations, error);
    }

    static new() {
        return new InMemoryLocationRepository([], false);
    }

    static withError() {
        return new InMemoryLocationRepository([], true);
    }

    insert(location: Location) {
        this.repo.insert(location);
    }

    fetchAll(): Promise<Result<Location[], string>> {
        return this.repo.fetchAll();
    }
}
